third_party/move/move-vm/runtime/src/storage/module_storage.rs

```diff
@@ -615,14 +601,42 @@ impl FunctionValueExtension for FunctionValueExtensionAdapter<'_> {
                 //      packed, all modules from all variants are loaded for value depth checks. In
                 //      case this change, we will be able to detect it as we pass traversal context
                 //      to check if this invariant has been violated.
-                // TODO(lazy-loading):
-                //   Use the context to check the invariant that all modules for layout are indeed
-                //   loaded. Will be coupled with native loader (later PRs).
-                let _ctx = UnvisitableModuleTraversalContext::new(traversal_context);
+                //
+                // As a result, we use a native loader here which will not visit any modules, and
+                // instead check that the accessed module is visited
+                let loader = NativeLoader::new(self.module_storage);
+                let layout_converter = LayoutConverter::new(&loader);
+
                 let captured_layouts = mask
                     .extract(fun.param_tys(), true)
                     .into_iter()
-                    .map(|t| ty_converter.type_to_type_layout(&instantiate(t)?))
+                    .map(|t| {
+                        let layout = layout_converter.type_to_type_layout_with_delayed_fields(
+                            // We do not expect gas to be metered here. The only thing that is
+                            // being checked is that all accessed modules have been loaded, hence
+                            // using unmetered gas meter and non-visitable context for checking.
+                            &mut UnmeteredGasMeter,
+                            &mut UnvisitableModuleTraversalContext::new(traversal_context),
```

Comment:
if we're confident that the invariant holds, here we can just use a Noop context?